semgrep-core/src/engine/Test_engine.ml

```diff
@@ -160,25 +184,70 @@ let make_tests ?(unit_testing = false) xs =
                    failwith
                      (spf "exn on %s (exn = %s)" file (Common.exn_to_s exn))
              in
-             res.profiling.rule_times
-             |> List.iter (fun rule_time ->
-                    if not (rule_time.RP.match_time >= 0.) then
-                      (* match_time could be 0.0 if the rule contains no pattern or if the
-                         rules are skipped. Otherwise it's positive. *)
-                      failwith
-                        (spf
-                           "invalid value for match time: %g (rule: %s, \
-                            target: %s)"
-                           rule_time.RP.match_time file target);
-                    if not (rule_time.RP.parse_time >= 0.) then
-                      (* same for parse time *)
-                      failwith
-                        (spf
-                           "invalid value for parse time: %g (rule: %s, \
-                            target: %s)"
-                           rule_time.RP.parse_time file target));
-
-             res.matches |> List.iter JSON_report.match_to_error;
+             let eres =
+               try
+                 Common.map
+                   (fun t ->
+                     let file = t.Input_to_core_t.path in
+                     let xlang = Xlang.of_string t.Input_to_core_t.language in
+                     let lazy_ast_and_errors =
+                       lazy
+                         (match xlang with
+                         | L (lang, _) ->
+                             let { Parse_target.ast; skipped_tokens; _ } =
+                               Parse_target.parse_and_resolve_name lang file
+                             in
+                             (ast, skipped_tokens)
+                         | LRegex
+                         | LGeneric ->
+                             assert false)
+                     in
+                     let xtarget =
+                       {
+                         Xtarget.file;
+                         xlang;
+                         lazy_content = lazy (Common.read_file file);
+                         lazy_ast_and_errors;
+                       }
+                     in
+
+                     let res =
+                       Match_rules.check
+                         ~match_hook:(fun _ _ -> ())
+                         ~timeout:0. ~timeout_threshold:0 (config, []) rules
+                         xtarget
+                     in
+                     Hashtbl.find_opt extract_result_map file
+                     |> Option.fold ~some:(fun f -> f res) ~none:res)
```

Comment:
Those Option.fold are ugly IMHO. They don't save you anything compared to just
match xxx with
| Some f -> f res
| None -> res